File: src/main/java/io/github/faecraft/heartshapedbox/math/FlexBox.java
package io.github.faecraft.heartshapedbox.math;

import net.minecraft.util.math.Vec3d;

// box with a flex
// how perplex
public class FlexBox {
    public final Quad[] quads = new Quad[6];
    
    // Not the actual center, but a good enough approximation for what its used for
    public final Vec3d roughCenter;
    
    public FlexBox(Vec3d a, Vec3d b, Vec3d c, Vec3d d, Vec3d e, Vec3d f, Vec3d g, Vec3d h) {
        quads[0] = new Quad(a, b, c, d);
        quads[1] = new Quad(e, f, g, h);
        quads[2] = new Quad(a, e, h, d);
        quads[3] = new Quad(b, f, e, a);
        quads[4] = new Quad(c, g, f, b);
        quads[5] = new Quad(d, h, g, c);
        
        roughCenter = a.add(h.subtract(a).multiply(0.5));
    }
    
    public FlexBox(Vec3d a, Vec3d b, Vec3d c, Vec3d d, double height) {
        Vec3d e = a.add(0, height, 0);
        Vec3d f = b.add(0, height, 0);
        Vec3d g = c.add(0, height, 0);
        Vec3d h = d.add(0, height, 0);
        
        quads[0] = new Quad(a, b, c, d);
        quads[1] = new Quad(e, f, g, h);
        quads[2] = new Quad(a, e, h, d);
        quads[3] = new Quad(b, f, e, a);
        quads[4] = new Quad(c, g, f, b);
        quads[5] = new Quad(d, h, g, c);
    
        roughCenter = a.add(h.subtract(a).multiply(0.5));
    }

    public FlexBox(Quad a, Quad b, Quad c, Quad d, Quad e, Quad f, Vec3d roughCenter) {
        quads[0] = a;
        quads[1] = b;
        quads[2] = c;
        quads[3] = d;
        quads[4] = e;
        quads[5] = f;

        this.roughCenter = roughCenter;
    }
    
    public static FlexBox zero() {
        return new FlexBox(Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO, Vec3d.ZERO);
    }
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/math/FlexBox.kt
package io.github.faecraft.heartshapedbox.math

import io.github.faecraft.heartshapedbox.util.FlexboxBaseBuilder
import io.github.faecraft.heartshapedbox.util.raise
import net.minecraft.util.math.Vec3d

// box with a flex
// how perplex
public data class FlexBox(
    val a: Quad,
    val b: Quad,
    val c: Quad,
    val d: Quad,
    val e: Quad,
    val f: Quad,
    val roughCenter: Vec3d
) {
    val quads: MutableList<Quad> = mutableListOf()

    init {
        quads.addAll(listOf(a, b, c, d, e, f))
    }

    @Suppress("MagicNumber")
    public constructor(a: Vec3d, b: Vec3d, c: Vec3d, d: Vec3d, e: Vec3d, f: Vec3d, g: Vec3d, h: Vec3d) : this(
        Quad(a, b, c, d),
        Quad(e, f, g, h),
        Quad(a, e, h, d),
        Quad(b, f, e, a),
        Quad(c, g, f, b),
        Quad(d, h, g, c),

        a.add(h.subtract(a).multiply(0.5))
    )

    @Suppress("MagicNumber")
    public constructor(a: Vec3d, b: Vec3d, c: Vec3d, d: Vec3d, height: Double) : this(
        Quad(a, b, c, d),
        Quad(a.raise(height), b.raise(height), c.raise(height), d.raise(height)),
        Quad(a, a.raise(height), d.raise(height), d),
        Quad(b, b.raise(height), a.raise(height), a),
        Quad(c, c.raise(height), b.raise(height), b),
        Quad(d, d.raise(height), c.raise(height), c),

        a.add(d.raise(height).subtract(a).multiply(0.5))
    )

    public constructor(builder: FlexboxBaseBuilder, height: Double) : this(
        builder.component1(),
        builder.component2(),
        builder.component3(),
        builder.component4(),
        height
    )

    public companion object {
        public val ZERO: FlexBox = FlexBox(
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO,
            Vec3d.ZERO
        )
    }
}
